### ISSUE TITLE
resolveCodeLens not aggressively cancelled when document changes

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#84185]
Issue Type: <b>Bug</b>

1. Enable TypeScript references code lenses
1. In the VS Code Code base, open `inspectKeybindings.ts`
1. On the line:

```
		return this._editorService.openEditor({ contents: this._keybindingService._dumpDebugInfoJSON(), options: { pinned: true } } as IUntitledTextResourceInput);

```

After `pinned: true }` type a comma then `re`

1. Delete and re-type `re` again

**Bug**
Suggestions take forever to come in

Computing the references in TypeScript can potentially be expensive. This is caused because the `resolveCodeLens` method is not being as aggressively canceled when the document changes as I expect it to be (and as I believe it used to be)

We can workaround this on the TS side but it seems like unexpected behavior

/cc @JacksonKearl 

VS Code version: Code - OSS Dev 1.41.0 (Commit unknown, Date unknown)
OS version: Darwin x64 18.7.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz (8 x 2200)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off<br>surface_control: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|11, 4, 3|
|Memory (System)|16.00GB (0.77GB free)|
|Process Argv|.|
|Screen Reader|no|
|VM|22%|
</details>

**Comment by mjbvz** [comment]
a8ad448cb0fd585512104640f3fc85f0a6270b16 Adds a TS specific fix. Whenever the document changes, we now aggressively cancel all existing code lens requests against the TS server for that document

**Comment by jrieken** [comment]
> canceled when the document changes as I expect it to be (and as I believe it used to be)

I believe nothing has change here - at least not as far as my recent memory goes... Cancellation happens after 250ms which isn't aggressive but should also not be perceived as overly slow. @mjbvz Are those 250ms the delay you were seeing?

**Comment by JacksonKearl** [comment]
@jrieken the delay is more like several seconds. It could be because we now provide references code lenses for object members, which adds a lot more references for TS to compute.

**Comment by jrieken** [comment]
hm... but the delay/debounce on the [render shouldn't be affected](https://github.com/microsoft/vscode/blob/3a34be0fd0ac863fa84a06f3522b6d96ee9dd9cf/src/vs/editor/contrib/codelens/codelensController.ts#L176-L177) by this... Does work happens on the extension host thread (not the tsserver-process) so that the cancellation doesn't come through until the event loop is free again?

**Comment by mjbvz** [comment]
Ok, so after more digging, I don't think is related to the debounce. Instead it seems some calls to `resolve` are not ever canceled when the document changes.

Here's the basic sequence I see on the extension host side:

```
Start resolve lenses #40-49 with document version:9
Start resolve lenses #50-62 with document version:9
Make an edit to file
Cancelled requests #50-62
Start resolve #63-70 with document version:13
Finished resolves #40-49 for document version 9.  Current document version: 13 
Finished resolves #63-70 for document version 13.  Current document version: 13 
```

The issue is that the first batch of resolve requests (40-49) are not cancelled when they should be

**Comment by jrieken** [comment]
Nice find and kudos for digging through that code 👏
